core/src/main/scala/org/apache/spark/deploy/master/MasterSource.scala

```diff
@@ -42,6 +42,6 @@ private[spark] class MasterSource(val master: Master) extends Source {
 
   // Gauge for waiting application numbers in cluster
   metricRegistry.register(MetricRegistry.name("waitingApps"), new Gauge[Int] {
-    override def getValue: Int = master.waitingApps.size
+    override def getValue: Int = master.apps.filter(_.state == ApplicationState.WAITING).size
```

Comment:
Do you always see these two values are different, from my understanding they should be same, perhaps there's some race condition which makes it different.